### ISSUE TITLE
Ctrl+C on stack test orphans child process

### ISSUE THREAD

**Issue by nh2** [commercialhaskell/stack#4416]
With some `hspec` code I'm currently writing, I noticed that when I run this

```haskell
import           Test.Hspec
import           Test.Hspec.QuickCheck (modifyMaxSize)
import qualified Test.QuickCheck as QC
import qualified Test.QuickCheck.Monadic as QCM
import           Test.QuickCheck.Instances.ByteString ()

--- ...

    describe "Identity" $ do
      modifyMaxSize (const 1000) $ it "compress and decompress arbitrary strings"$
        QC.property $ \bsList -> QCM.monadicIO $ do
          QCM.run $ hPutStrLn stderr (show $ sum $ map BS.length bsList) -- extremely slow
          QCM.assert True
```

and it runs and I press Ctrl+C, it keeps polluting my shell prompt.

This suggests that my Ctrl+C kills `stack` but not the test it launches.

**Comment by nh2** [comment]
I can confirm that when running the test executable directly, and pressing Ctrl+C, it ignores the first one and only dies to the second.

**Comment by snoyberg** [comment]
I'm tempted to close as wontfix. Test processes are always capable of escaping via the double-fork approach, for example, so I'm not convinced this is something Stack should care about.

**Comment by nh2** [comment]
> Test processes are always capable of escaping via the double-fork approach

That is correct, but I don't think that is what's happening here, because (to my knowledge) the test is not double-forking.

I suspect what is happening is that Stack sends the process SIGTERM, and considers its job done and itself allowed to terminate, even when it can still observe that it has a living child process, and it still has the option to send SIGKILL, which would solve the issue. I conclude that from the fact that pressing Ctrl+C from on the test executable directly has no effect but does _not_ return to the terminal.

It is also possible that I am wrong, and there's an intermediate executable between stack and the test executable that I'm not aware of, which actually does terminate and Stack reaps its exit code.

**Comment by snoyberg** [comment]
I agree with your analysis. My point is: I'm not convinced this is a case that Stack _should_ be trying to solve better. We know that test executables can be written that will escape this approach anyway, so there's no security-sensitive approach Stack can be following.

**Comment by nh2** [comment]
Ah, my point is not about security, only about the developer annoyance that stuff continues to run despite telling it to stop.

I think that if Stack's children double-fork, there's nothing it can do and it should exit.

But I think if Stack still has children, then it should not terminate until they are gone. Even in the case where it cannot get them to die (e.g. if they are hung in the kernel and don't respond to even SIGKILL), it would be better for Stack to not terminate, so that the user knows that there's something hanging (as the next call to `stack test` will likely fail if there's something still running in the background).

In other words, it'd be OK for Stack to exit if a child orphaned itself, but there should be no case where Stack _creates_ orphans by exiting despite knowledge of own children.

**Comment by snoyberg** [comment]
Alright, you've convinced (and nerd sniped) me.

**Comment by snoyberg** [comment]
I wanted to provide a graceful shutdown, but from what I can tell, if I use `threadDelay` in the Stack executable _after_ SIGTERM has been sent to it, the thread immediately exits. Do you have experience with this @nh2? As it stands, it looks like all I can do is send a SIGTERM followed immediately by a SIGKILL.
